### ISSUE TITLE
set_length of GrooveJoint2D doesn't change its length

### ISSUE THREAD

**Issue by stephanbogner** [godotengine/godot#85144]
### Godot version

v4.1.3.stable.official [f06b6836a]

### System information

Apple Ventura 13.5, Forward+

## Setting the `length` of a `GrooveJoint2D` doesn't work. The simulated length doesn't actually get updated. It keeps the initially set length.

---

## My guess
It seems the joint's parameters [are not actually changed](https://github.com/godotengine/godot/blob/ac29e8e3fed8a8c8d74a267a60eee75b979bb994/scene/2d/joint_2d.cpp#L436), unlike in e.g. [PinJoint2D::set_softness](https://github.com/godotengine/godot/blob/ac29e8e3fed8a8c8d74a267a60eee75b979bb994/scene/2d/joint_2d.cpp#L294).

[There is](https://github.com/godotengine/godot/blob/master/servers/physics_2d/godot_physics_server_2d.cpp) a `pin_joint_set_param` and a `damped_spring_joint_set_param`, but no `groove_joint_set_param`. If I can find the time I will try to add this function and integrate it into `GrooveJoint2D::set_length`.

## Test setup:
- The left rigid body is connected to a groove joint with a length of 200px ...
- ... and the right one to a groove joint with the length of 50px **but** a script sets the length also to 200px

```
func _ready():
	$Right/GrooveJoint.set_length(200)
```

<img width="995" alt="Screenshot 2023-11-20 at 17 50 44" src="https://github.com/godotengine/godot/assets/7897006/9bfacba9-d4c2-4410-83ae-37364dfd8e38">

### Expected result
Both rigid bodies should behave the exact same.

### Result
The visualisation of the right groove joint is updated correctly (yellow line), but it doesn't fall lower than the initially specified 50px.

<img width="688" alt="Screenshot 2023-11-20 at 17 58 00" src="https://github.com/godotengine/godot/assets/7897006/ad3910ec-053e-47c6-a6b5-73b4ebe5aa97">

---

PS: I seem to become the guy that opens all the 2D physics bugs 😅 

### Steps to reproduce

1. Create a `GrooveJoint2D` connected to two bodies
2. Create a script that changes the `length` of the GrooveJoint (neither `length` nor `set_length` does work)
3. Watch the simulation play

### Minimal reproduction project

[groove-joint-bug.zip](https://github.com/godotengine/godot/files/13417593/groove-joint-bug.zip)

**Comment by AThousandShips** [comment]
The required code is indeed missing, will take a look at fixing this

**Comment by stephanbogner** [comment]
@AThousandShips Wow that was a fast reply. Thanks for investigating!

Good to know regarding the reconfiguration. I don't plan to update it dynamically, only to set it once in the beginning according to an exported variable. So at least for me that bug fix would be sufficient and great 🙂

**Comment by jsjtxietian** [comment]
> it does however require the joint to be reconfigured each time as there's no way to update it

I do feel it's a little stricter. Maybe we can do something about it.

**Comment by stephanbogner** [comment]
@jsjtxietian Thanks for looking into it!
(This also made me compile Godot for the first time 🎉 )

### A) Setting it once works really nicely!
Underwhelming screenshot but it works as expected 😅
<img width="688" alt="Screenshot 2023-11-24 at 10 30 06" src="https://github.com/godotengine/godot/assets/7897006/1ebdde73-0dd0-4756-b88b-bbd720ac1cab">

### B) However making it longer later (e.g. after 1 second) makes it behave unexpectedly:
- The ball stops down way further than it should and doesn't go up as far (it appears that the joint has been shifted downwards)
- If I apply an impulse it swings which the unchanged groove joint does not (but it doesn't swing when I set it in A)

https://github.com/godotengine/godot/assets/7897006/cdfdc061-ade5-4ada-aebf-d312ef44e199

### Code replacement for my original example:
```
extends Node2D

func _ready():
	$Right/GrooveJoint.length = 200
	await get_tree().create_timer(2).timeout
	$Left/RigidBody.apply_central_impulse(Vector2(100, -1000))
	$Right/GrooveJoint.set_length(205)
	await get_tree().create_timer(1).timeout
	$Right/RigidBody.apply_central_impulse(Vector2(100, -1000))
```

---

### Summary: It seems to solve my problem of setting length initially, but it would require docs to clarify that it can't be changed during runtime (if that is not an easy fix).

**Comment by jsjtxietian** [comment]
Thank you for your test and congrates to your first compilation of Godot! 

> * The ball stops down way further than it should and doesn't go up as far (it appears that the joint has been shifted downwards)
> * If I apply an impulse it swings which the unchanged groove joint does not (but it doesn't swing when I set it in A)

Yes that's indeed the problem, other joint2d's usually have a interface for them to directly change the specific para that are set by user, while GrooveJoint2D do not.

**Comment by stephanbogner** [comment]
@jsjtxietian Thanks for clarifying. For me your changes would already be a huge improvement :)

FYI: I did some more tests (that I [posted in the pull request](https://github.com/godotengine/godot/pull/85162)) and it looks great.

**Comment by RomanKoziolek** [comment]
Hi all, thank you for the contributions!
I will need to modify the GrooveJoint2D's `length` during runtime in order to control sail line length in a sailing game. Do you know of any workaround that I could use to get around this problem?

> The required code is indeed missing, it does however require the joint to be reconfigured each time as there's no way to update it

Regarding comment above, I may not fully understand the difference between `reconfiguration` and `update`. Could you clarify?

**Comment by AThousandShips** [comment]
"Update" here means say to the physics engine "set the length to this", whereas "reconfigure" means "create a new joint with this configuration" 🙂

**Comment by RomanKoziolek** [comment]
thank you @AThousandShips.

If I understand this correctly, when this issue is resolved, one could change the length of the GrooveJoint2D by creating new 
 joint with desired length and initial offset and reapply to objects while removing the old joint?

**Comment by AThousandShips** [comment]
No, that's not it, what I'm saying is that any solution would have to work around it, see the linked PR, this won't change anything from a user perspective, just an implementation detail 🙂

**Comment by RomanKoziolek** [comment]
ok, you're right, code changes won't solve the update in runtime problem.
I think I'll create an issue and try figuring out how to work around it.
Thanks for the help!

**Comment by AThousandShips** [comment]
An issue for what? This issue here is solved by the PR, in runtime

**Comment by sbsmirnoff** [comment]
Hey guys,

Great job keeping stuff rolling!
Any idea when this PR will be merged and released?

Thanks!
